includes/class-newspack-ads-gam.php

```diff
@@ -0,0 +1,386 @@
+<?php
+/**
+ * Newspack Ads GAM management
+ *
+ * @package Newspack
+ */
+
+use Google\AdsApi\Common\Configuration;
+use Google\AdsApi\AdManager\AdManagerServices;
+use Google\AdsApi\AdManager\AdManagerSessionBuilder;
+use Google\AdsApi\AdManager\Util\v202102\StatementBuilder;
+use Google\AdsApi\AdManager\v202102\ServiceFactory;
+use Google\AdsApi\AdManager\v202102\ArchiveAdUnits as ArchiveAdUnitsAction;
+use Google\AdsApi\AdManager\v202102\AdUnit;
+use Google\AdsApi\AdManager\v202102\AdUnitSize;
+use Google\AdsApi\AdManager\v202102\AdUnitTargetWindow;
+use Google\AdsApi\AdManager\v202102\EnvironmentType;
+use Google\AdsApi\AdManager\v202102\Size;
+
+require_once NEWSPACK_ADS_COMPOSER_ABSPATH . 'autoload.php';
+
+/**
+ * Newspack Ads GAM Management
+ */
+class Newspack_Ads_GAM {
+	// https://developers.google.com/ad-manager/api/soap_xml: An arbitrary string name identifying your application. This will be shown in Google's log files.
+	const GAM_APP_NAME_FOR_LOGS = 'Newspack';
+
+	/**
+	 * Codes of networks that the user has access to.
+	 *
+	 * @var Network[]
+	 */
+	private static $networks = [];
+
+	/**
+	 * Reusable GAM session.
+	 *
+	 * @var AdManagerSession
+	 */
+	private static $session = null;
+
+	/**
+	 * Get OAuth2 credentials.
+	 *
+	 * @throws \Exception If the user is not authenticated.
+	 * @return object OAuth2 credentials.
+	 */
+	private static function get_google_oauth2_credentials() {
+		if ( class_exists( 'Newspack\Google_Services_Connection' ) ) {
+			$oauth2_credentials = \Newspack\Google_Services_Connection::get_oauth2_credentials();
+			if ( false === $oauth2_credentials ) {
+				throw new \Exception( __( 'Please authenticate Newspack with Google.', 'newspack-ads' ), 1 );
+			}
+			return $oauth2_credentials;
+		} else {
+			return new \Exception( __( 'Please activate the Newspack Plugin.', 'newspack-ads' ), 1 );
+		}
+	}
+
+	/**
+	 * Get GAM networks the authenticated user has access to.
+	 *
+	 * @return Network[] Array of networks.
+	 */
+	private static function get_gam_networks() {
+		if ( empty( self::$networks ) ) {
+			// Create a configuration and session to get the network codes.
+			$config = new Configuration(
+				[
+					'AD_MANAGER' => [
+						'networkCode'     => '-', // Provide non-empty network code to pass validation.
+						'applicationName' => self::GAM_APP_NAME_FOR_LOGS,
+					],
+				]
+			);
+
+			$oauth2_credentials = self::get_google_oauth2_credentials();
+			$session            = ( new AdManagerSessionBuilder() )->from( $config )->withOAuth2Credential( $oauth2_credentials )->build();
```

Comment:
Does this have to be a separate session from the one used to interact with the API (created on L129)?

Reply:
Yes, this is a bit of the GAM PHP API weirdness - you need a network id in a session to get ad units (and everything else), but to get the network id you need another session 🤷‍♀️